[performance] ensure `causal_mask` is created directly on device

jeffra opened the pull request:
# What does this PR do?

@tjruwase and @tohtana discovered that causal_mask is currently being created on CPU then moved to GPU during the forward pass of OPT (and we think other models). This appears to be causing a significant performance degradation on multi-gpu environments due to parallel host to device copies going on. It's not 100% clear to us why this is so bad but here is what we observe before and after this patch:

Before this patch w. OPT-125m on x8 A100s:
<img width="649" alt="image" src="https://user-images.githubusercontent.com/645595/227668447-bf6840dd-bbc4-4520-8a9f-33f046eeb4c2.png">

After the patch:
<img width="628" alt="image" src="https://user-images.githubusercontent.com/645595/227668475-6ed2f1ca-d18a-4776-862d-4be499f62f39.png">

These numbers were gathered from a modified version of https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py but turning on `wall_clock_breakdown: true` in our deepspeed config.

One major complication we see in accepting this PR is that the two functions being modified are copied across lots of different models and the `make fix-copies` script doesn't seem to address all of them correctly across both `_make_causal_mask` and `_prepare_decoder_attention_mask`

## Who can review?

Tagging @sgugger and @stas00 to help triage to the right people

jeffra committed:
ensure causal_mask is created directly on device

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22378). All of your documentation changes will be reflected on that endpoint.

ydshieh commented:
cc @thomasw21 @NouamaneTazi since both of you are experts on this kind of things - to see if you have any general opinion and/or if you would like to review this PR too.

ydshieh commented:
@jeffra Would it possible for you (and/or @tjruwase and @tohtana) to provide your script that finds/measures/profiles the running time for this issue 🙏 . It would be super helpful for us to dive into internally too.

NouamaneTazi commented on src/transformers/models/opt/modeling_opt.py:
```diff
@@ -64,18 +64,20 @@
     bsz, tgt_len = input_ids_shape
-    mask = torch.full((tgt_len, tgt_len), torch.tensor(torch.finfo(dtype).min))
-    mask_cond = torch.arange(mask.size(-1))
+    mask = torch.full((tgt_len, tgt_len), torch.tensor(torch.finfo(dtype).min, device=device), device=device)
+    mask_cond = torch.arange(mask.size(-1), device=device)
```
```suggestion
    mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min, device=device)
    mask_cond = torch.arange(mask.size(-1), device=device)
```
It seems that `torch.tensor(torch.finfo(torch.float32).min, device="cuda")` requires CPU-GPU sync

NouamaneTazi replied:
Otherwise if a tensor is needed, we can also do
```
torch.ones(1, device="cuda") * torch.finfo(torch.float32).min # no sync
```

NouamaneTazi replied:
I just tried my suggestion:`mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min, device=device)` 
and it does work on "cpu". I'm not sure why it wouldn't work? 🤔

thomasw21 replied:
From the pytorch doc (https://pytorch.org/docs/stable/generated/torch.full.html), `torch.full` allows for scalar type for `fill_value`. So I think what @NouamaneTazi is trying to convey is that you don't need to put it first cast `-inf` to a tensor type and then fill, you could just fill with `-inf`

nmntz reviewed (commented):
I'm a big supporter of removing CPU-GPU syncs, so I would very like see this merged! ⚡️

sgugger reviewed (approved):
LGTM, thanks a lot for the fix! Note that the same modification needs to be applied to BART (since OPT copies from BART) in order for all quality checks to pass.

sgugger commented:
I think just copying the same fix to BART and then applying `make fix-copies` is simple enough for this PR. Dealing with functions that are not copies or are named differently can indeed be done in followup PRs.

jeffra committed:
Merge branch 'huggingface:main' into main

jeffra committed:
add copy tag to opt, update bart implementation

jeffra commented:
Ok, i've updated the BART implementation and attempted again to get `make fix-copies` to work for me but I think I might be doing something wrong. Some of the original issues I saw are now fixed on other models (e.g., https://github.com/huggingface/transformers/pull/22382 adds a `# Copied from` tag for llama). However, I am still seeing issues i think coming from the fix-up scripts getting confused with the function signature change of `_make_causal_mask`.

Can someone try `make fix-copies` on their side with this? You should be able to push to my branch.

sgugger commented:
Ah yes, `make fix-copies` does not change the signature of the function so that is indeed something to edit manually. If it's too much work I can try to push this to your branch tomorrow.

jeffra commented:
> Ah yes, `make fix-copies` does not change the signature of the function so that is indeed something to edit manually. If it's too much work I can try to push this to your branch tomorrow.

Sounds good, I might have some time this afternoon for this. Otherwise feel free to do it :)

jeffra committed:
add device to all _make_causal_mask copies

jeffra committed:
formatting fixes

jeffra commented:
Okay all the models should be fixed now, `make fixup` is clear on my local tests.

jeffra committed:
more manual fixes due to unlinked versions of _prepare_decoder_attention_mask

sgugger reviewed (approved):
Thanks for copy-pasting the signature changes manually!
